lms/djangoapps/grades/tasks.py

```diff
@@ -315,14 +316,28 @@ def _update_subsection_grades(
     student = User.objects.get(id=user_id)
     store = modulestore()
     with store.bulk_operations(course_key):
-        course_structure = get_course_blocks(student, store.make_course_usage_key(course_key))
+        course_usage_key = store.make_course_usage_key(course_key)
+        course_structure = get_course_blocks(student, course_usage_key)
         subsections_to_update = course_structure.get_transformer_block_field(
             scored_block_usage_key,
             GradesTransformer,
             'subsections',
             set(),
         )
 
+        # When access is restricted and course blocks are cached without the restricted blocks,
+        # we will update the cache from the store and get the course blocks.
+        if not subsections_to_update:
+            update_course_in_cache(course_usage_key.course_key)
```

Comment:
I think this could be a long-running process and we should see the impact in ways like:

1. Can we use a cache updating method that just updates the cache for `scored_block_usage_key` and not the whole course?
2. How long with it generally take on larger courses?
3. I see a newer version of this method [update_course_in_cache_v2](https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/content/block_structure/tasks.py#L41) - would it be better to shift towards that?

Reply:
> Can we use a cache updating method that just updates the cache for scored_block_usage_key and not the whole course?

I tried to look more into it and initially thought there might be a way to do it for the scored block. It looked like `BlockStructureManager` might support but it didn't.
There is another way to get the block from the store directly but that does not cover all the use cases where `recalculate_subsection_grade_v3` is used. Maybe someone having more in-depth knowledge could suggest a better approach if any.

> How long with it generally take on larger courses?

I have tested it for 3 courses locally and here is the breakdown:
Demonstration Course: 5 Sections, 11 Subsections, 45 Units: **1.5 Secs**
Course X: 40 Sections, 96 Subections, 185 Units: **10 Secs**
Course Y: 13 Sections, 33 Subsections, 289 Units: **13 Secs**

Also, we should note that this update won't be a frequent one. This approach is used at multiple places where we fetch the blocks of a course.

> I see a newer version of this method [update_course_in_cache_v2](https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/content/block_structure/tasks.py#L41) - would it be better to shift towards that?

We are already doing the update inside a celery task and delaying it to another task will not serve the purpose.

Reply:
@asadali145 I think  problem with using `update_course_in_cache` is it would not update blockstructure cache if `STORAGE_BACKING_FOR_CACHE` switch is enabled and course published version is same as BlockStructureModel version. See implementation of [is_up_to_date](https://github.com/openedx/edx-platform/blob/547f7febcc79a6a748d31b0fb97c5ef4027e625d/openedx/core/djangoapps/content/block_structure/store.py#L118) for details how it makes update decision.

I think we can clear the blockstructure cache using `clear_course_from_cache` and then raise exception SUBSECTION_NOT_FOUND(need add it in one of the known [retry exceptions](https://github.com/openedx/edx-platform/blob/547f7febcc79a6a748d31b0fb97c5ef4027e625d/lms/djangoapps/grades/tasks.py#L250)). This would retry the `recalculate_subsection_grade_v3` task.